tsdb/importer/importer.go

```diff
@@ -0,0 +1,583 @@
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	level.Debug(logger).Log("msg", "reading existing block metadata to infer time ranges")
+	dbMint, dbMaxt, blockTimes, err := getDbTimes(dbPath)
+	if err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "creating new block skeletons based on existing blocks")
+	blockMetas := constructBlockMetadata(dbMint, dbMaxt, blockTimes)
+
+	level.Info(logger).Log("msg", "importing input data and aligning with existing DB")
+	if err = writeSamples(r, tmpDbDir, dbMint, dbMaxt, blockMetas, maxSamplesInMemory, logger); err != nil {
+		return err
+	}
+
+	level.Info(logger).Log("msg", "merging fully overlapping blocks")
+	if err = mergeBlocks(tmpDbDir, blockMetas, logger); err != nil {
+		return err
+	}
+
+	level.Debug(logger).Log("msg", "copying newly created blocks from temp location to actual DB location")
+	if err = fileutil.CopyDirs(tmpDbDir, dbPath); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// writeSamples parses each metric sample, and writes the samples to the correspondingly aligned block.
+// It uses the block indexes assigned to each sample, and block meta info, gathered in collectSampleInformation().
+func writeSamples(r io.Reader, dbDir string, dbMint, dbMaxt int64, metas []*newBlockMeta, maxSamplesInMemory int, logger log.Logger) error {
+	blockTimes := make([]blockTimestampPair, 0)
+	for _, meta := range metas {
+		blockTimes = append(blockTimes, blockTimestampPair{
+			start: meta.mint,
+			end:   meta.maxt,
+		})
+	}
+
+	blocks := getEmptyBlocks(len(metas))
+	currentPassCount := 0
+
+	// TODO: Maybe use bufio.Reader.Readline instead?
+	// https://stackoverflow.com/questions/21124327/how-to-read-a-text-file-line-by-line-in-go-when-some-lines-are-long-enough-to-ca
+	// Use a streaming approach to avoid loading too much data at once.
+	scanner := bufio.NewScanner(r)
+	buf := new(bytes.Buffer)
+	scanner.Split(sampleStreamer(buf))
+
+	for scanner.Scan() {
+		if currentPassCount == 0 {
+			blocks = getEmptyBlocks(len(metas))
+		}
+
+		encSample := scanner.Bytes()
+		decBuf := bytes.NewBuffer(encSample)
+		sample := tsdb.MetricSample{}
+		if err := gob.NewDecoder(decBuf).Decode(&sample); err != nil {
+			level.Error(logger).Log("msg", "failed to decode current entry returned by file scanner", "err", err)
+			return err
+		}
+
+		// Find what block does this sample belong to.
+		var blockIndex int
+		if len(metas) == 1 || sample.Timestamp < dbMint {
+			blockIndex = 0
+		} else if sample.Timestamp >= dbMaxt {
+			blockIndex = len(metas) - 1
+		} else {
+			blockIndex = getBlockIndex(sample.Timestamp, blockTimes)
+		}
+
+		if blockIndex == -1 {
+			return NoBlockMetaFoundError
+		}
+
+		meta := metas[blockIndex]
+		meta.mint = value.MinInt64(meta.mint, sample.Timestamp)
+		meta.maxt = value.MaxInt64(meta.maxt, sample.Timestamp)
+		meta.count += 1
+
+		blocks[blockIndex] = append(blocks[blockIndex], &sample)
```

Comment:
don't understand why you need the binSamples step?

when you have the blocks time ranges constructed from an existing db or from the default time ranges you can just add the samples directly to the struct

Something like:
```
map[index]newBlocks{ 
	count
	mint
	maxt
	[]samples
}

range samples {
	idx:= getBlockIndex(sample) 
	newBlocks[idx]samples=append(newBlocks[idx]samples,sample)
	newBlocks[idx]count++
	if newBlocks[idx]count > ...{
		flush(newBlocks[idx])
	}
}
```

Reply:
The binning only incurs an overhead for samples outside DB time limits - for the aligned case it just creates a single bin.

Reply:
If the existing time ranges are 100-200, 200-400 and you you have a sample 50 or 500 you can create more time ranges on the fly. 
In this case for 500 add a new block that starts from 500 and end at 500+ the range of the last block in the array and for 50 add a new block from 50 minus the first block in the array, not going below zero.

Reply:
Using the binning approach makes things a lot easier for me to understand, but I do see your point about perhaps adding child blocks or whatever on the fly. I'm going to have to think about how to do this.

Reply:
I've been thinking about adding blocks on the fly (for samples that are outside DB limits), and I think that speculatively guessing time ranges for them will lead to more blocks that could possibly need to be compacted.

in the current setup, I use binning to create blocks with the default duration of 2h, which I think is pretty straightforward! 

I also ran a few benches, and with the test cases I have, with a `max-samples-in-mem` of 10,000, and even in the worst case the binning step took less than 0.5 ms, with most cases taking much lesser than that.